### ISSUE TITLE
Pinned dependencies prevent datalab from being installed in existing environments

### ISSUE THREAD

**Issue by MaximilianR** [googledatalab/datalab#1641]
Because of the pinned dependencies, installing datalab into an existing environments downgrades every dependency of datalab's (which are a significant number)

I realize there are benefits of pinned dependencies - particularly if you're the only top-level application that's going to be running. But this does prevent datalab from being installed in an environment with other applications, or in environments where code depends on later versions of libraries (e.g. the pandas version is fairly old here)

FWIW datalab is the only major library I've seen that does this; though I do understand its context is particular

**Comment by nikhilk** [comment]
We've seen cases where certain combinations of libraries cause issues, so we have to find a set of versions that can co-exist, rather than blindly picking up what is latest at any given point in time.

As a start I think we should allow for at least some range of dependencies. Example, instead of `scikit-learn==0.18.2`, we should express this as `scikit-learn>=0.18,<0.19`. That seems like a fix we should consider.

Second, we should look at which libraries are in most need of an update. Are there any particular ones you're seeing issues that you can help identify/prioritize?

Third, I think we'll want to version our datalab container images, so a given version represents a certain set of libraries and their versions, and we can have newer versions of Datalab that pick up newer versions of dependencies.

Finally as a totally ugly workaround you may be able to use the `--no-deps` flag with `pip install` when installing datalab and then take on manual installation of dependencies ... though in doing so, you're also taking on responsibility of a working environment.

**Comment by MaximilianR** [comment]
> We've seen cases where certain combinations of libraries cause issues, so we have to find a set of versions that can co-exist, rather than blindly picking up what is latest at any given point in time.

Yes, I do appreciate that. On the other hand, there are many libraries that face a similar predicament, and manage (though at cost - they need to run a test matrix, respond quickly when issues come up, etc)

> Are there any particular ones you're seeing issues that you can help identify/prioritize?

Pandas & numpy; though we're probably more heavy users of the new features than others are

> Finally as a totally ugly workaround you may be able to use the --no-deps flag with pip install

Great point, thanks

Finally - thank you for building such an impressive library. The reason we care about this is that we'd really love to start using your hard work!
